File: scripts/radio-offset-diagnostic.mjs
#!/usr/bin/env node
/**
 * Precise diagnostic: measure radio vs checkbox child-element offset
 * against the container div to find the source of the ~2-3px radio misalignment.
 */
import { resolve, dirname, join } from 'path';
import { fileURLToPath } from 'url';
import { createRequire } from 'module';

const __dirname = dirname(fileURLToPath(import.meta.url));
const appDir = resolve(__dirname, '../app');
const require = createRequire(join(appDir, 'package.json'));
const { chromium } = require('playwright');

const APP = 'http://localhost:3000';

const browser = await chromium.launch({ headless: true, args: ['--no-sandbox'] });
const ctx = await browser.newContext({ viewport: { width: 1800, height: 1200 }, deviceScaleFactor: 3 });
const page = await ctx.newPage();

// Auth
const email = `diag-${Date.now()}@test.dev`;
const pw = 'diagCheck1234';
await fetch(`${APP}/api/auth/register`, {
  method: 'POST', headers: { 'Content-Type': 'application/json' },
  body: JSON.stringify({ email, password: pw }),
});
await page.goto(`${APP}/login`, { waitUntil: 'domcontentloaded', timeout: 60000 });
await page.waitForTimeout(2000);
await page.fill('input#email', email);
await page.fill('input#password', pw);
await page.click('button[type="submit"]');
for (let i = 0; i < 30; i++) { await page.waitForTimeout(500); if (!page.url().includes('/login')) break; }

// Create form
await page.goto(`${APP}/new`, { waitUntil: 'domcontentloaded', timeout: 60000 });
await page.waitForTimeout(3000);
await page.click('button:has-text("Create Form")');
for (let i = 0; i < 30; i++) { await page.waitForTimeout(500); if (page.url().includes('/identification/section1')) break; }
const subId = page.url().split('/')[3];
console.log(`Form: ${subId}`);

async function setupPdfLayout() {
  const btn = page.locator('button:has-text("PDF Layout")');
  if (await btn.count() > 0) { await btn.click(); await page.waitForTimeout(2000); }
  await page.evaluate(() => {
    const s = document.querySelector('input[type="range"]');
    if (s) {
      Object.getOwnPropertyDescriptor(HTMLInputElement.prototype, 'value').set.call(s, '100');
      s.dispatchEvent(new Event('input', { bubbles: true }));
      s.dispatchEvent(new Event('change', { bubbles: true }));
    }
  });
  const sa = page.locator('label:has-text("Show All") input[type="checkbox"]');
  if (await sa.count() > 0 && !(await sa.isChecked())) await sa.check();
  const sf = page.locator('label:has-text("Show fields") input[type="checkbox"]');
  if (await sf.count() > 0 && !(await sf.isChecked())) await sf.check();
  await page.waitForTimeout(3000);
  await page.evaluate(async () => {
    const imgs = document.querySelectorAll('.relative.shadow-md img');
    await Promise.all(Array.from(imgs).map(img =>
      img.complete ? Promise.resolve() : new Promise(r => { img.onload = r; img.onerror = r; })
    ));
  });
  await page.waitForTimeout(1000);
}

// ========================= Section 9 =========================
console.log('\n=== Section 9 — Citizenship ===');
await page.goto(`${APP}/${subId}/citizenship/section9`, { waitUntil: 'domcontentloaded', timeout: 60000 });
await page.waitForTimeout(5000);
try { await page.waitForFunction(() => window.__JOTAI_STORE__ && window.__FIELD_VALUE_ATOM_FAMILY__, { timeout: 15000 }); } catch {}
await setupPdfLayout();

const info = await page.evaluate(() => {
  const results = [];
  const fields = document.querySelectorAll('[data-field-key]');
  const frames = document.querySelectorAll('.relative.shadow-md');
  if (frames.length === 0) return { error: 'No frames' };

  const frameRect = frames[0].getBoundingClientRect();
  const img = frames[0].querySelector('img');
  const imgRect = img ? img.getBoundingClientRect() : null;

  const frameInfo = {
    frameX: frameRect.x, frameY: frameRect.y,
    frameW: frameRect.width, frameH: frameRect.height,
    imgX: imgRect?.x, imgY: imgRect?.y,
    imgW: imgRect?.width, imgH: imgRect?.height,
    imgNatW: img?.naturalWidth, imgNatH: img?.naturalHeight,
  };

  for (const el of fields) {
    const key = el.getAttribute('data-field-key');
    const rect = el.getBoundingClientRect();
    // Only fields on first frame
    if (rect.y < frameRect.y - 5 || rect.y > frameRect.y + frameRect.height + 5) continue;

    const isRadio = key.includes('radio');
    const child = el.firstElementChild;
    const childRect = child ? child.getBoundingClientRect() : null;
    const childCS = child ? getComputedStyle(child) : null;

    results.push({
      key: key.substring(0, 70),
      isRadio,
      childTag: child?.tagName || 'none',
      // Container position relative to frame
      relX: rect.x - frameRect.x,
      relY: rect.y - frameRect.y,
      cW: rect.width, cH: rect.height,
      // Child offset from container
      chOffX: childRect ? childRect.x - rect.x : null,
      chOffY: childRect ? childRect.y - rect.y : null,
      chW: childRect?.width, chH: childRect?.height,
      // Child computed styles
      chPadT: childCS?.paddingTop,
      chPadL: childCS?.paddingLeft,
      chMarT: childCS?.marginTop,
      chMarL: childCS?.marginLeft,
      chBorT: childCS?.borderTopWidth,
      chBorL: childCS?.borderLeftWidth,
      chBoxSz: childCS?.boxSizing,
      chDisp: childCS?.display,
    });
  }

  return { frameInfo, fields: results };
});

console.log('\n--- Frame Info ---');
const fi = info.frameInfo;
console.log(`Frame: (${fi.frameX.toFixed(1)}, ${fi.frameY.toFixed(1)}) ${fi.frameW.toFixed(1)}x${fi.frameH.toFixed(1)}`);
console.log(`Image: (${fi.imgX?.toFixed(1)}, ${fi.imgY?.toFixed(1)}) ${fi.imgW?.toFixed(1)}x${fi.imgH?.toFixed(1)}`);
console.log(`Image natural: ${fi.imgNatW}x${fi.imgNatH}`);
console.log(`Frame-Image offset: dx=${((fi.imgX || 0) - fi.frameX).toFixed(3)}, dy=${((fi.imgY || 0) - fi.frameY).toFixed(3)}`);
console.log(`Scale factor: ${(fi.frameW / 612).toFixed(6)}`);

console.log('\n--- Field positions on first page ---');
const header = `${'Key'.padEnd(55)} ${'Type'.padEnd(6)} ${'Tag'.padEnd(8)} ${'relX'.padStart(8)} ${'relY'.padStart(8)} ${'cW'.padStart(6)} ${'cH'.padStart(6)} ${'chOffY'.padStart(8)} ${'chW'.padStart(6)} ${'chH'.padStart(6)} ${'padT'.padStart(6)} ${'marT'.padStart(6)} ${'borT'.padStart(6)} ${'boxSz'.padStart(12)} ${'disp'.padStart(14)}`;
console.log(header);
for (const f of info.fields.slice(0, 25)) {
  const t = f.isRadio ? 'RADIO' : 'OTHER';
  console.log(
    `${f.key.padEnd(55)} ${t.padEnd(6)} ${f.childTag.padEnd(8)} ${f.relX.toFixed(2).padStart(8)} ${f.relY.toFixed(2).padStart(8)} ${f.cW.toFixed(1).padStart(6)} ${f.cH.toFixed(1).padStart(6)} ${(f.chOffY ?? 0).toFixed(3).padStart(8)} ${(f.chW ?? 0).toFixed(1).padStart(6)} ${(f.chH ?? 0).toFixed(1).padStart(6)} ${(f.chPadT || '').padStart(6)} ${(f.chMarT || '').padStart(6)} ${(f.chBorT || '').padStart(6)} ${(f.chBoxSz || '').padStart(12)} ${(f.chDisp || '').padStart(14)}`
  );
}

// Check if radio children have nonzero offset
const radioFields = info.fields.filter(f => f.isRadio);
const otherFields = info.fields.filter(f => !f.isRadio);
console.log(`\n--- Summary ---`);
console.log(`Radio fields: ${radioFields.length}`);
console.log(`Other fields: ${otherFields.length}`);

if (radioFields.length > 0) {
  const avgRadioOffY = radioFields.reduce((s, f) => s + (f.chOffY || 0), 0) / radioFields.length;
  const avgRadioOffX = radioFields.reduce((s, f) => s + (f.chOffX || 0), 0) / radioFields.length;
  console.log(`Avg radio child offset: X=${avgRadioOffX.toFixed(3)}, Y=${avgRadioOffY.toFixed(3)}`);
}
if (otherFields.length > 0) {
  const avgOtherOffY = otherFields.reduce((s, f) => s + (f.chOffY || 0), 0) / otherFields.length;
  const avgOtherOffX = otherFields.reduce((s, f) => s + (f.chOffX || 0), 0) / otherFields.length;
  console.log(`Avg other child offset: X=${avgOtherOffX.toFixed(3)}, Y=${avgOtherOffY.toFixed(3)}`);
}

// Take tight crops of the first radio and first other field
if (radioFields.length > 0) {
  const r = radioFields[0];
  const el = page.locator(`[data-field-key="${r.key}"]`).first();
  await el.scrollIntoViewIfNeeded();
  await page.waitForTimeout(300);
  const box = await el.boundingBox();
  if (box) {
    await page.screenshot({
      path: '/tmp/diag_s9_radio.png',
      clip: { x: Math.max(0, box.x - 15), y: Math.max(0, box.y - 15), width: box.width + 30, height: box.height + 30 },
    });
    console.log(`\nRadio crop: /tmp/diag_s9_radio.png (${box.width.toFixed(1)}x${box.height.toFixed(1)})`);
  }
}

// Radio pair crop
if (radioFields.length >= 2) {
  const el0 = page.locator(`[data-field-key="${radioFields[0].key}"]`).first();
  const el1 = page.locator(`[data-field-key="${radioFields[1].key}"]`).first();
  await el0.scrollIntoViewIfNeeded();
  await page.waitForTimeout(200);
  const b0 = await el0.boundingBox();
  const b1 = await el1.boundingBox();
  if (b0 && b1) {
    const mx = Math.min(b0.x, b1.x);
    const my = Math.min(b0.y, b1.y);
    const mxr = Math.max(b0.x + b0.width, b1.x + b1.width);
    const myr = Math.max(b0.y + b0.height, b1.y + b1.height);
    await page.screenshot({
      path: '/tmp/diag_s9_radio_pair.png',
      clip: { x: Math.max(0, mx - 20), y: Math.max(0, my - 15), width: mxr - mx + 40, height: myr - my + 30 },
    });
    console.log(`Radio pair crop: /tmp/diag_s9_radio_pair.png`);
  }
}

if (otherFields.length > 0) {
  const o = otherFields[0];
  const el = page.locator(`[data-field-key="${o.key}"]`).first();
  await el.scrollIntoViewIfNeeded();
  await page.waitForTimeout(300);
  const box = await el.boundingBox();
  if (box) {
    await page.screenshot({
      path: '/tmp/diag_s9_other.png',
      clip: { x: Math.max(0, box.x - 15), y: Math.max(0, box.y - 15), width: box.width + 30, height: box.height + 30 },
    });
    console.log(`Other crop: /tmp/diag_s9_other.png (${box.width.toFixed(1)}x${box.height.toFixed(1)})`);
  }
}

await browser.close();
console.log('\nDone!');

File: scripts/radio-offset-diagnostic.py
#!/usr/bin/env python3
"""
Precise diagnostic: measure radio vs checkbox alignment offset
against the PDF background image in the live browser.
"""
from playwright.sync_api import sync_playwright
import json, time

APP = "http://localhost:3000"

with sync_playwright() as p:
    browser = p.chromium.launch(headless=True, args=["--no-sandbox"])
    ctx = browser.new_context(viewport={"width": 1800, "height": 1200}, device_scale_factor=3)
    page = ctx.new_page()

    # Auth
    email = f"diag-{int(time.time())}@test.dev"
    pw = "diagCheck1234"
    import urllib.request
    req = urllib.request.Request(
        f"{APP}/api/auth/register",
        data=json.dumps({"email": email, "password": pw}).encode(),
        headers={"Content-Type": "application/json"},
        method="POST",
    )
    urllib.request.urlopen(req, timeout=10)

    page.goto(f"{APP}/login", wait_until="domcontentloaded", timeout=60000)
    page.wait_for_timeout(2000)
    page.fill("input#email", email)
    page.fill("input#password", pw)
    page.click('button[type="submit"]')
    for _ in range(30):
        page.wait_for_timeout(500)
        if "/login" not in page.url:
            break

    # Create form
    page.goto(f"{APP}/new", wait_until="domcontentloaded", timeout=60000)
    page.wait_for_timeout(3000)
    page.click('button:has-text("Create Form")')
    for _ in range(30):
        page.wait_for_timeout(500)
        if "/identification/section1" in page.url:
            break
    sub_id = page.url.split("/")[3]
    print(f"Form: {sub_id}")

    def setup_pdf_layout():
        btn = page.locator('button:has-text("PDF Layout")')
        if btn.count() > 0:
            btn.click()
            page.wait_for_timeout(2000)
        # Set opacity 100%
        page.evaluate("""() => {
            const s = document.querySelector('input[type="range"]');
            if (s) {
                Object.getOwnPropertyDescriptor(HTMLInputElement.prototype, 'value').set.call(s, '100');
                s.dispatchEvent(new Event('input', { bubbles: true }));
                s.dispatchEvent(new Event('change', { bubbles: true }));
            }
        }""")
        # Show All + Show fields
        sa = page.locator('label:has-text("Show All") input[type="checkbox"]')
        if sa.count() > 0 and not sa.is_checked():
            sa.check()
        sf = page.locator('label:has-text("Show fields") input[type="checkbox"]')
        if sf.count() > 0 and not sf.is_checked():
            sf.check()
        page.wait_for_timeout(3000)
        # Wait for images
        page.evaluate("""async () => {
            const imgs = document.querySelectorAll('.relative.shadow-md img');
            await Promise.all(Array.from(imgs).map(img =>
                img.complete ? Promise.resolve() : new Promise(r => { img.onload = r; img.onerror = r; })
            ));
        }""")
        page.wait_for_timeout(1000)

    # Go to section 5 (page 5 0-indexed = 4, which has both radios and checkboxes)
    page.goto(f"{APP}/{sub_id}/identification/section5", wait_until="domcontentloaded", timeout=60000)
    page.wait_for_timeout(5000)
    try:
        page.wait_for_function(
            "() => window.__JOTAI_STORE__ && window.__FIELD_VALUE_ATOM_FAMILY__",
            timeout=15000,
        )
    except:
        pass
    setup_pdf_layout()

    # Diagnostic: inject colored debug markers directly on the page
    # Measure the positions of radio and checkbox elements vs the image
    info = page.evaluate("""() => {
        const results = [];

        // Get all field elements with data-field-key
        const fields = document.querySelectorAll('[data-field-key]');

        // Get the first PDF page frame
        const frames = document.querySelectorAll('.relative.shadow-md');
        if (frames.length === 0) return { error: 'No PDF frames found' };

        const frameRect = frames[0].getBoundingClientRect();
        const img = frames[0].querySelector('img');
        const imgRect = img ? img.getBoundingClientRect() : null;

        const frameInfo = {
            frameX: frameRect.x, frameY: frameRect.y,
            frameW: frameRect.width, frameH: frameRect.height,
            imgX: imgRect?.x, imgY: imgRect?.y,
            imgW: imgRect?.width, imgH: imgRect?.height,
            imgNaturalW: img?.naturalWidth, imgNaturalH: img?.naturalHeight,
        };

        // Get radio and checkbox elements on this frame
        for (const el of fields) {
            const key = el.getAttribute('data-field-key');
            const rect = el.getBoundingClientRect();
            const style = el.getAttribute('style') || '';

            // Check if it's within the first frame
            if (rect.y < frameRect.y || rect.y > frameRect.y + frameRect.height) continue;

            const isRadio = key.includes('radio');
            const child = el.firstElementChild;
            const childTag = child ? child.tagName : 'none';
            const childRect = child ? child.getBoundingClientRect() : null;

            // Compute position relative to frame
            const relX = rect.x - frameRect.x;
            const relY = rect.y - frameRect.y;

            results.push({
                key: key.substring(0, 60),
                isRadio,
                childTag,
                // Container (absolute div) position
                containerX: rect.x,
                containerY: rect.y,
                containerW: rect.width,
                containerH: rect.height,
                // Position relative to frame
                relX, relY,
                // CSS style values
                cssLeft: parseFloat(style.match(/left:\\s*([\\d.]+)px/)?.[1] || '0'),
                cssTop: parseFloat(style.match(/top:\\s*([\\d.]+)px/)?.[1] || '0'),
                // Child element position
                childX: childRect?.x,
                childY: childRect?.y,
                childW: childRect?.width,
                childH: childRect?.height,
                // Offset between container and child
                childOffsetX: childRect ? childRect.x - rect.x : null,
                childOffsetY: childRect ? childRect.y - rect.y : null,
            });
        }

        return { frameInfo, fields: results.slice(0, 30) };
    }""")

    print(f"\n=== Frame Info ===")
    fi = info.get("frameInfo", {})
    print(f"Frame: ({fi.get('frameX'):.1f}, {fi.get('frameY'):.1f}) {fi.get('frameW'):.1f}x{fi.get('frameH'):.1f}")
    print(f"Image: ({fi.get('imgX'):.1f}, {fi.get('imgY'):.1f}) {fi.get('imgW'):.1f}x{fi.get('imgH'):.1f}")
    print(f"Image natural: {fi.get('imgNaturalW')}x{fi.get('imgNaturalH')}")
    print(f"Frame-Image offset: dx={fi.get('imgX', 0) - fi.get('frameX', 0):.2f}, dy={fi.get('imgY', 0) - fi.get('frameY', 0):.2f}")

    print(f"\n=== Field Positions (first 30) ===")
    print(f"{'Key':40} {'Type':8} {'Tag':8} {'relX':>8} {'relY':>8} {'cssL':>8} {'cssT':>8} {'cW':>6} {'cH':>6} {'childOffX':>9} {'childOffY':>9}")
    for f in info.get("fields", []):
        ftype = "RADIO" if f["isRadio"] else "OTHER"
        print(f"{f['key']:40} {ftype:8} {f['childTag']:8} {f['relX']:8.2f} {f['relY']:8.2f} {f['cssLeft']:8.2f} {f['cssTop']:8.2f} {f['containerW']:6.1f} {f['containerH']:6.1f} {f.get('childOffsetX', 0) or 0:9.2f} {f.get('childOffsetY', 0) or 0:9.2f}")

    # Take crops
    # Find a radio element and take tight crop
    radios = [f for f in info.get("fields", []) if f["isRadio"]]
    others = [f for f in info.get("fields", []) if not f["isRadio"]]

    if radios:
        r = radios[0]
        pad = 15
        page.screenshot(
            path="/tmp/diag_radio_crop.png",
            clip={
                "x": max(0, r["containerX"] - pad),
                "y": max(0, r["containerY"] - pad),
                "width": r["containerW"] + pad * 2,
                "height": r["containerH"] + pad * 2,
            },
        )
        print(f"\nRadio crop saved: /tmp/diag_radio_crop.png")

    if others:
        o = others[0]
        pad = 15
        page.screenshot(
            path="/tmp/diag_other_crop.png",
            clip={
                "x": max(0, o["containerX"] - pad),
                "y": max(0, o["containerY"] - pad),
                "width": o["containerW"] + pad * 2,
                "height": o["containerH"] + pad * 2,
            },
        )
        print(f"Other crop saved: /tmp/diag_other_crop.png")

    # Take a big crop showing both types
    if radios and others:
        all_fields = radios[:3] + others[:3]
        min_x = min(f["containerX"] for f in all_fields)
        min_y = min(f["containerY"] for f in all_fields)
        max_x = max(f["containerX"] + f["containerW"] for f in all_fields)
        max_y = max(f["containerY"] + f["containerH"] for f in all_fields)
        page.screenshot(
            path="/tmp/diag_combo_crop.png",
            clip={
                "x": max(0, min_x - 30),
                "y": max(0, min_y - 20),
                "width": max_x - min_x + 60,
                "height": max_y - min_y + 40,
            },
        )
        print(f"Combo crop saved: /tmp/diag_combo_crop.png")

    # Now let's try section 9 which had the most visible issues
    print("\n\n=== Section 9 ===")
    page.goto(f"{APP}/{sub_id}/citizenship/section9", wait_until="domcontentloaded", timeout=60000)
    page.wait_for_timeout(5000)
    try:
        page.wait_for_function(
            "() => window.__JOTAI_STORE__ && window.__FIELD_VALUE_ATOM_FAMILY__",
            timeout=15000,
        )
    except:
        pass
    setup_pdf_layout()

    s9_info = page.evaluate("""() => {
        const results = [];
        const fields = document.querySelectorAll('[data-field-key]');
        const frames = document.querySelectorAll('.relative.shadow-md');
        if (frames.length === 0) return { error: 'No frames' };

        const frameRect = frames[0].getBoundingClientRect();

        for (const el of fields) {
            const key = el.getAttribute('data-field-key');
            const rect = el.getBoundingClientRect();
            if (rect.y < frameRect.y || rect.y > frameRect.y + frameRect.height) continue;

            const isRadio = key.includes('radio');
            const child = el.firstElementChild;
            const childRect = child ? child.getBoundingClientRect() : null;

            results.push({
                key: key.substring(0, 60),
                isRadio,
                childTag: child ? child.tagName : 'none',
                relX: rect.x - frameRect.x,
                relY: rect.y - frameRect.y,
                containerW: rect.width,
                containerH: rect.height,
                childOffsetX: childRect ? childRect.x - rect.x : null,
                childOffsetY: childRect ? childRect.y - rect.y : null,
                childW: childRect?.width,
                childH: childRect?.height,
            });
        }
        return { frameW: frameRect.width, frameH: frameRect.height, fields: results };
    }""")

    print(f"Frame: {s9_info.get('frameW', 0):.1f}x{s9_info.get('frameH', 0):.1f}")
    print(f"\n{'Key':50} {'Type':6} {'Tag':8} {'relX':>7} {'relY':>7} {'cW':>5} {'cH':>5} {'chOffX':>7} {'chOffY':>7} {'chW':>5} {'chH':>5}")
    for f in s9_info.get("fields", [])[:20]:
        ftype = "RADIO" if f["isRadio"] else "OTHER"
        print(f"{f['key']:50} {ftype:6} {f['childTag']:8} {f['relX']:7.2f} {f['relY']:7.2f} {f['containerW']:5.1f} {f['containerH']:5.1f} {f.get('childOffsetX', 0) or 0:7.2f} {f.get('childOffsetY', 0) or 0:7.2f} {f.get('childW', 0) or 0:5.1f} {f.get('childH', 0) or 0:5.1f}")

    # Take s9 radio crops at 3x
    s9_radios = [f for f in s9_info.get("fields", []) if f["isRadio"]]
    if s9_radios:
        for i, r in enumerate(s9_radios[:4]):
            el = page.locator(f'[data-field-key="{r["key"]}"]').first
            el.scroll_into_view_if_needed()
            page.wait_for_timeout(200)
            box = el.bounding_box()
            if box:
                page.screenshot(
                    path=f"/tmp/s9_radio_{i}.png",
                    clip={
                        "x": max(0, box["x"] - 12),
                        "y": max(0, box["y"] - 12),
                        "width": box["width"] + 24,
                        "height": box["height"] + 24,
                    },
                )
        print(f"\nSaved s9 radio crops")

    browser.close()
    print("\nDone!")
